### PR TITLE
Add blessings to stats editor

### PR DESCRIPTION
**Pull request by soarqin**
Add DLC blessings to stats editor, which requires https://github.com/veeenu/practice-tool-core/pull/19 to be merged first.

### CODE REVIEW

File: practice-tool/src/widgets/character_stats.rs [hunk 1]
```diff
@@ -1,19 +1,34 @@
...
+            if let Some(blessings) = self.blessings.as_mut() {
+                has_blessings = true;
+                b = blessings;
+            } else {
+                has_blessings = false;
+                unsafe {
+                    static mut BLESSINGS_NONE: CharacterBlessings = CharacterBlessings { scadutree: 0, revered_spirit_ash: 0 };
+                    b = BLESSINGS_NONE.borrow_mut();
+                }
+            }
```
Comment (veeenu) [inline practice-tool/src/widgets/character_stats.rs:29]: The semantics here are a bit confusing, significantly complicate the method, and use `unsafe` as a wiorkaround which we should avoid at all costs.

The following pseudocode [chaining](https://doc.rust-lang.org/std/iter/struct.Chain.html) iterators would be much cleaner:

```rs
self.stats.as_mut().map(|s| {
    let base_data = [ /* ...the result of the previous closure */ ].into_iter();

    if let Some(b) = self.blessings.as_mut() {
        base_data.chain([
            Datum::byte("Scadutree Blessing", &mut b.scadutree, 0, 20),
            Datum::byte("Revered Spirit Ash", &mut b.revered_spirit_ash, 0, 20),
        ].into_iter())
    } else {
        base_data
    }
```

File: practice-tool/src/widgets/character_stats.rs [hunk 2]
```diff
@@ -22,34 +37,52 @@ impl Stats for CharacterStatsEdit {
                 Datum::int("Intelligence", &mut s.intelligence, 1, 99),
                 Datum::int("Faith", &mut s.faith, 1, 99),
                 Datum::int("Arcane", &mut s.arcane, 1, 99),
-                Datum::int("Souls", &mut s.runes, 1, i32::MAX),
-            ]
-            .into_iter()
-        })
+                Datum::int("Souls", &mut s.runes, 0, i32::MAX),
+                Datum::byte("Scadutree Blessing", &mut b.scadutree, 0, if has_blessings { 20 } else { -1 }),
+                Datum::byte("Revered Spirit Ash", &mut b.revered_spirit_ash, 0, if has_blessings { 10 } else { -1 }),
```
Comment (veeenu) [inline practice-tool/src/widgets/character_stats.rs:42]: I see now how this relates to my comment in https://github.com/veeenu/practice-tool-core/pull/19. We should _never_ use sentinel values to represent states as that has a lot of potential for confusion. Addressing my comment above should make this check unnecessary as the iterator is pre-filtered.

**Review by veeenu — CHANGES_REQUESTED** [review]
This is a great feature to have, thank you so much!

File: practice-tool/src/widgets/character_stats.rs (hunk 2 above)
Comment (soarqin) [reply practice-tool/src/widgets/character_stats.rs:42]: The thing is that I wanna hide blessings values from stats editor when the version is older than DLC release (or make them disabled at least), coz memory addresses would be invalid here.
If not using unsafe block, I have to do something more complex to implement global temporary variables(e.g. add a member field to CharacterStats)

File: practice-tool/src/widgets/character_stats.rs (hunk 1 above)
Comment (soarqin) [reply practice-tool/src/widgets/character_stats.rs:29]: This would imply compiler error due to different lengths in returned Iterrator from 2 branches

**Comment by soarqin** [comment]
Turned out to make it working with your suggestions, and removed `max` value check in practice-core.
I have force pushed new codes to both repo, please recheck them again.

File: practice-tool/src/widgets/character_stats.rs (hunk 1 above)
Comment (veeenu) [reply practice-tool/src/widgets/character_stats.rs]: Oh, right, that makes sense yeah.

File: practice-tool/src/widgets/character_stats.rs [hunk 3]
```diff
@@ -5,15 +5,17 @@ use practice_tool_core::widgets::Widget;
...
     stats: Option<CharacterStats>,
+    blessings: Option<CharacterBlessings>,
 }
 
 impl Stats for CharacterStatsEdit {
     fn data(&mut self) -> Option<impl Iterator<Item = Datum>> {
         self.stats.as_mut().map(|s| {
-            [
-                Datum::int("Level", &mut s.level, 1, i32::MAX),
+            let mut stats_data = vec![
```
Comment (veeenu) [inline practice-tool/src/widgets/character_stats.rs:17]: This is on the hot path, I'd prefer avoiding allocations and using a buffer preallocated into the struct and `.drain(..)`ed, but let's leave it like this for now. I need to make a pass over the whole codebase to ensure no allocations are made per frame anyway so I can deal with that myself.

**Review by veeenu — APPROVED** [review]
Looks good, thank you so much, this is great 🙌

File: practice-tool/src/widgets/character_stats.rs (hunk 3 above)
Comment (soarqin) [reply practice-tool/src/widgets/character_stats.rs:17]: The allocations are only executed on each subwindow's popup.
Also, you can just make 2 pre-allocated vectors as caches, if time-save for the allocations are really needed

Comment (veeenu) [reply practice-tool/src/widgets/character_stats.rs:17]: IIUC it should be executed on every frame, as `imgui` is not retained. The input widgets are recreated at every frame and you can modify them while the popup is open.

So far per-frame allocations haven't been a huge deal honestly, but one of the design goals of the tool is to take as little frame time as possible to allow the game to maximise its usage for itself.

**Pull request merged**